### PR TITLE
[WIP] Better Logger

### PR DESCRIPTION
**Pull request by ezrast**
See https://github.com/crystal-lang/crystal/issues/5874

Still needs specs and documentation, but see https://gist.github.com/ezrast/37c3377c172cd225efe0ed40f4da1e59 for example usage.

### CODE REVIEW

**Comment by RX14** [comment]
I like the interface but the implementation is a little on the complex side. I'd rather have a singleton root logger instead of `extend Base`. I don't think `Logger.log` should work either, you should need to `.get` a logger or use `Logger::BASE` (not recommended).

The `HierarchyFilter` *could* work on longest-substring match, which avoids assuming that `::` is used as a separator. This can be done now by creating an array of the filter rules, ordered by match string length, and then iterating that. Later, replace with a radix tree.

This is a really good starting point though!

**Comment by RX14** [comment]
Perhaps also just have the `Filter` on the `Emitter` in this design, with the global `Logger::ROOT` (`BASE` was a bad name) controlling which emitter is used, and the emitter handling filtering.


So perhaps:

```cr
class Logger
  property backend : Logger | Emitter
end

class Emitter
  property filter : Filter
end
```

This design allows having a "root logger" which isn't `Logger::BASE`, meaning there can be many trees of loggers, but usually only one with the root being `Logger::BASE`. The root of the logger tree references the emitter which controls where the logs go, how they're formatted, and if they're filtered.

File: src/logger/entry.cr [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    getter line_number : Int32
+    getter filename : String
+    @message : String | Proc(String)
+
+    def initialize(@message, @severity, @component, @time, @line_number, @filename)
+    end
+
+    def message : String
+      if (msg = @message).is_a? Proc(String)
+        return @message = msg.call
```
Comment (Sija) [inline src/logger/entry.cr]: redundant `return`, ditto below.

**Comment by ezrast** [comment]
> The `HierarchyFilter` _could_ work on longest-substring match, which avoids assuming that `::` is used as a separator. This can be done now by creating an array of the filter rules, ordered by match string length, and then iterating that. Later, replace with a radix tree.

If we don't assume a separator then `Animals::Anteater` will inherit from `Animals::Ant`. I think we want to avoid that?

> Perhaps (not sure) also just have the `Filter` on the `Emitter` in this design, with the global `Logger::ROOT` (`BASE` was a bad name) controlling which emitter is used, and the emitter handling filtering.
> 
> So perhaps:
> 
> ```crystal
> class Logger
>   getter component : String
>   property backend : Logger | Emitter
> end
> 
> class Emitter
>   property filter : Filter
> end
> ```

The issue with this is that it makes the filter difficult to access from the outside - changing the default log level becomes `Logger::ROOT.backend.as?(Emitter).try &.filter= Logger::DEBUG`, which isn't great. If you want to avoid having `nil` filters on all the forwarders, I think the way to do it is to make `Logger.get` return a `Forwarder` object that doesn't inherit from `Logger` at all. Then `#debug` and friends live on `Forwarder` instead of, or in addition to, on `Logger`. I'm not sure whether that actually simplifies anything, but I'll play with it.

> This design allows having a "root logger" which isn't `Logger::BASE`, meaning there can be many trees of loggers, but usually only one with the root being `Logger::BASE`. The root of the logger tree references the emitter which controls where the logs go, how they're formatted, and if they're filtered.

Whoops, I had meant to move `get` from a class method on `Logger` to an instance method on `Logger::Base`. Then one can create forwarders for any arbitrary logger.

**Comment by RX14** [comment]
@ezrast if you add a `filter=` on `Logger` which asserts that this is the root logger and then sets the filter then it's just `Logger::ROOT.filter = `

**Comment by j8r** [comment]
@ezrast Note you haven't set the a mail linked with your github's account in your `.gitconfig`

**Comment by ezrast** [comment]
@j8r Good catch! That explains why Github thinks I've never committed before.

@RX14 That still seems like fighting the type system. If we don't want components to have their own filters we should give them a type that doesn't have filters. Does this change do what you want in that regard? https://github.com/ezrast/crystal/pull/3/files

**Comment by RX14** [comment]
@ezrast if you go that route then i'd rather just have `Logger` and `RootLogger < Logger` and only `RootLogger` has the emitter *and* filter.

But I'll wait for other people on the core team to chime in since they'll have their own opinions.

**Comment by straight-shoota** [comment]
This `Component` type is what I already had in mind in https://github.com/crystal-lang/crystal/issues/5874#issuecomment-376679699 where the directly used object would just be a shallow wrapper with a component path and a pointer to the logger instance. Since it's immutable, it could even be a struct and make calling `Logger.get` pretty cheap.

However, this implies a very centralized architecture with every log message routing through one main logger instance. It can get pretty complex to apply filters to specific component paths.

So, I'm not sure that's the best solution.

Looking at Java's logging API, every `Logger` instance has a `filter` and a list of `handlers` as well as an optional reference to a `parent` logger. Log messages are published to the local handlers and the handlers of all parent loggers (this propagation can be disabled).

**Comment by RX14** [comment]
I tend to think that copying what java does is the best. There have been about 5-10 java logging libraries over the years, and they're very well optimized by now for production usecases.

**Comment by RX14** [comment]
I've checked through way too many levels of abstraction to confirm that in log4j2, each logger has a filter, a set of appenders (handlers) and then forwards to it's parent. I'd probably just copy that for now, although it does require a registry of logger instances to look up parents, but that's required anyway to reconfigure loggers at runtime.

File: src/logger.cr [hunk 2]
```diff
@@ -1,178 +1,56 @@
...
+    abstract def emitter : EmitterType?
+
+    def log(entry : Entry)
+      case ff = filter
+      when Filter, Proc(Entry, Bool)
+        return unless ff.call(entry)
+      when Severity
+        return unless entry.severity >= ff
+      end
+      emitter.try &.call(entry)
```
Comment (bcardiff) [inline src/logger.cr]: It was hard to follow how the `Logger.class` that is a `Logger::Base` ends up chained as the default `emitter`.

File: src/logger.cr [hunk 3]
```diff
@@ -1,178 +1,56 @@
...
-    end
-
-    # Logs the message as returned from the given block if the logger's current severity
-    # is lower or equal to `{{name.id}}`. The block is not run if the severity is higher.
-    # *progname* overrides a default progname set in this logger.
-    def {{name.id.downcase}}(progname = nil)
-      log(Severity::{{name.id}}, progname) { yield }
-    end
+  {% for level in Severity.constants %}
+    {{ level }} = Severity::{{ level }}
```
[the remark this answers is not in the document]
Comment (ezrast) [reply src/logger.cr]: It's a holdover from the current implementation (https://github.com/crystal-lang/crystal/blob/master/src/logger.cr#L130). Of course it's less useful now with #6074. I'll take this out.

File: src/logger/entry.cr [hunk 4]
```diff
@@ -0,0 +1,23 @@
+require "./severity"
+
+class Logger
+  class Entry
```
[the remark this answers is not in the document]
Comment (ezrast) [reply src/logger/entry.cr]: `Entry` mutates itself so that message strings are evaluated at most once in the case of calls like `logger.debug{ do_expensive_thing.to_s }`. IMO this is a good semantic to preserve.

File: src/logger.cr (hunk 3 above)
[the remark this answers is not in the document]
Comment (bcardiff) [reply src/logger.cr]: Oh, ok. Let's remove it later.

File: src/logger/entry.cr (hunk 4 above)
[the remark this answers is not in the document]
Comment (straight-shoota) [reply src/logger/entry.cr]: I'm not sure that's worth it. Log messages are usually not very expensive and then it's much better to avoid allocations.

**Comment by RX14** [comment]
I think we still have architecture issues to discuss before nitpicking this PR.

**Comment by ezrast** [comment]
@RX14, as a very general response to your Java comment: in my experience Java things tend to be a pain to deal with operationally, and one of the many reasons is that their components write things to lots of different log files, including duplicating entries between log files, all loosely controlled by a configuration file according to a scheme that is largely inscrutable to non-Java programmers. So I'm in favor of pushing configuration into the root logger instance. I don't want to encourage components to write to their own log files "just in case" and then also push logs downstream to be written again somewhere else.

That said, changing `Logger` to hold an array of `Emitter`s seems like a benign change, and after that (plus making `get` an instance method) what we'll have isn't that different from what you and @straight-shoota describe: calling `Logger#get` will return another logger that forwards to its parent by default but can have a filter and arbitrary emitters added to it.

The main thing missing is a registry, which I don't understand the use case for. Global configuration should go in the root logger; why make every component's logger be available globally?

**Comment by ezrast** [comment]
> I definitely would like to see a module the user can be referenced to emit events and a factory method. And decouple the story of filtering in another module/interface. That should help the description to the user.

Can you clarify what you mean by this? `Emitter` and `Filter` are already their own modules that will be documented individually.

File: src/logger/entry.cr (hunk 4 above)
[the remark this answers is not in the document]
Comment (bcardiff) [reply src/logger/entry.cr]: I agree with @straight-shoota . It is tempting the idea of lazy building the logger message, but not at the cost of stressing the GC.

**Comment by RX14** [comment]
@ezrast okay, yes, root logger with an array of `Emitter`s, each emitter receiving every log and having separate filtering settings is fine. No registry or parenting arrangement required.

Also the naming is iffy. I'd propose:
- `Logger::Manager` - Holds the array of `Collector`s
- `Logger::Collector` - Handles the core logic of filtering and serializing logs
- `Logger::Filter`
- `Logger` itself is a struct containing a `String` context and a reference to the `Logger::Manager` instance
- `Logger::DEFAULT_MANAGER`, a default `Manager`
- `Logger::Manager#get(String)`, returns a `Logger`
- `Logger.get` - forwards to `Logger::DEFAULT_MANAGER.get`

Thoughts?

**Comment by RX14** [comment]
@straight-shoota because if loggers have state, they need to be stored in a registry so that they can be configured using a central API. That complicates the whole design.

**Comment by straight-shoota** [comment]
Hm, yeah. But now `Logger::Manager` is kind of a registry. It at least needs to associate filters to component paths. I'm not favouring any of both approaches, just thinking about it...

**Comment by ezrast** [comment]
The current PR has no registry, distinguishes between components hierarchically in the root logger if you want it to (for which the entire filter class is 26 lines of code and not complex at all), and still has fat loggers at the leaves just in case you want to do something wacky. I'm skeptical that it could get much less complicated.

**Comment by RX14** [comment]
@straight-shoota except in this design the "registry" (really just parsing the component) is handled by `Collector`s, and collectors can be implemented by shards. This moves almost all the xomplexity outside the stdlib and allows flexibility. The stdlib will still provide good solutions for deployments with only a few rules (see the API for `Logger::HierarchyFilter` and compare that to a registry)

File: src/logger.cr [hunk 5]
```diff
@@ -7,47 +7,59 @@ class Logger
...
-    {% for level in Severity.constants %}
-      def {{ level.downcase.id }}(*, time = Time.now, line_number = __LINE__, filename = __FILE__, &message : -> String)
-        log Entry.new(message, Severity::{{ level }}, component, time, line_number, filename)
-      end
+  def log(entry : Entry)
+    case ff = filter
+    when Filter, Proc(Entry, Bool)
+      return unless ff.call(entry)
+    when Severity
+      return unless entry.severity >= ff
```
Comment (j8r) [inline src/logger.cr]: You mean `return if entry.severity < ff`

File: src/logger.cr [hunk 6]
```diff
@@ -7,32 +7,29 @@ class Logger
...
-      case ff = filter
-      when Filter, Proc(Entry, Bool)
-        return unless ff.call(entry)
-      when Severity
-        return unless entry.severity >= ff
-      end
-      emitters.each &.call(entry)
+  property component : String
+  property filter : FilterType?
+  property emitters : Array(EmitterType)
```
Comment (straight-shoota) [inline src/logger.cr]: Why did the ivar accessors became properties? I don't think they need to be mutable. `getter` should be enough.

**Comment by ezrast** [comment]
I made some of the less-controversial (I think; feel free to prove me wrong) requested changes and updated https://gist.github.com/ezrast/37c3377c172cd225efe0ed40f4da1e59 accordingly.

Outstanding discussion topics include:
* Make `Logger#get` return a thin forwarder instance (removes a footgun but also some flexibility; makes the implementation slightly more complex) or keep it as a fat logger that we just pretend is thin (simplest solution but potentially confusing)?
* Filter-per-emitter or filter-per-logger?
* What to name things? (should be deferred until the architecture is nailed down)

**Comment by straight-shoota** [comment]
* I'm still not sure about the thin wrappers, but I figure we can leave it as is for now, as it is a sound solution. Once we get some experience with real world usage of the loggers, and determine that thin wrappers would provide an improvement, this can always be revisited.
* I'd say per logger. That makes it easy to share the same emitters across different logger instances. If you need different filter settings, you can just create another logger instance and forward messages to it. An alternative would be a `FilterEmitter` which wraps an `Emitter` and filters messages before relaying them. But it's not much different from another `Logger` plus `Forwarder`.
* The current naming actually looks quite nice.

File: src/logger/entry.cr (hunk 4 above)
[the remark this answers is not in the document]
Comment (ezrast) [reply src/logger/entry.cr]: I think lazy evaluation will save allocations in real-world scenarios, but putting them back in should be a non-breaking change, so I've removed them and will re-raise the issue in a later ticket if I can generate some meaningful benchmarks.

[the remark this answers is not in the document]
Comment (straight-shoota) [reply src/logger/entry.cr]: Yeah, lazy evaluation itself really isn't the issue. It's memoizing the result...

File: src/logger.cr (hunk 5 above)
Comment (j8r) [reply src/logger.cr]: "if X is inferior than Y" is simpler to understand than "unless X is superior or equal than Y", my opinion.

File: src/logger.cr (hunk 6 above)
Comment (ezrast) [reply src/logger.cr]: I made `filter` a property since `Logger::ROOT.filter = ...` needs to be possible, and then brought the rest along for consistency. But I agree immutability would be better in other cases. I think that's an argument in favor of making `Logger.get` return a class that is thinner than `Logger`.

`component` can be made immutable, in any case. `emitters` doesn't really matter since the array will always be mutable.

Comment (straight-shoota) [reply src/logger.cr]: Ah yes, I see the last commit. Yeah, when we're not using thin wrappers, `filter` should probably be mutable. But `component` and `emitters` do not.

**Comment by RX14** [comment]
Ultimately, what i'm trying to say is that the data flow graph for my proposal is this:

```
 Logger --\                      /--> Logger::Collector -------> ...                                            
          |                      |           ^                                                                  
          |                      |           |                                                                  
 Logger --|--> Logger::Manager --|           |                                                                  
          |                      |           v                                                                  
          |                      |    Logger::Filter                                                            
 Logger --/                      |                                                                            
                                 |                                                                          
                                 \--> Logger::Collector -------> ...                                                                  
                                             ^                                                                                       
                                             |                                                                                       
                                             |                                                                                       
                                             v                                                                                       
                                      Logger::Filter                                                                                 
```

whereas, the same graph for this PR is spaghetti

**Comment by ezrast** [comment]
> Logger needs to be stateless and immutable, because otherwise you need to keep a registry of loggers to make filters and appenders configurable by the application.

@RX14, a registry isn't in the question at this point. It's "thin loggers; central configuration" vs "fat loggers; central configuration". In practice, the current PR's data flow is like this --
```
 Logger (leaf) --> Logger::Forwarder --\                    /--> Logger::Emitter --> ...
                                       |                    |
                                       |                    |
 Logger (leaf) --> Logger::Forwarder --|--> Logger (root) --|--> Logger::Emitter --> ...
                                       |          ^         |
                                       |          |         |
 Logger (leaf) --> Logger::Forwarder --/          |         \--> Logger::Emitter --> ...
                                                  v
                                            Logger::Filter
```
-- which has the *potential* to turn into spaghetti if people go adding filters and emitters to the leaf loggers, but the docs should make it clear that that's not the intended use case. Otherwise, the only effective difference between your graph and mine is where the filters go, which is a separate question.

I am not actually against the idea of giving leaf loggers their own immutable class, I just think you're overstating the difference between the two setups.

File: src/logger.cr [hunk 7]
```diff
@@ -1,178 +1,68 @@
...
-
-    # A warning
-    WARN
-
-    # A handleable error condition
-    ERROR
+require "logger/entry"
+require "logger/filter"
+require "logger/emitter"
+require "logger/severity"
```
[the remark this answers is not in the document]
Comment (ezrast) [reply src/logger.cr]: If you prefer, sure.

[the remark this answers is not in the document]
Comment (straight-shoota) [reply src/logger.cr]: IMHO individual includes are preferable. They show exactly what gets included and where. And you can apply an order which can be of significance.

**Comment by straight-shoota** [comment]
One important difference between both data flow diagrams isn't shown. In RX14's design, `Logger.get("Foo")` will always return essentially and identical logger whereas in the current state it will always return a new instance and when you modify it, with say a new filter, other instances will not be affected. This makes it the duty of the application to ensure it shares its logger.

**Comment by RX14** [comment]
@ezrast that's what i meant by "Logger needs to be stateless and immutable, because otherwise you need to keep a registry of loggers". If Loggers hold configuration data, applications then need to be able to access the loggers of their libraries through `.get`, which requires a registry. If loggers are immutable, then the config is *entirely* in the `Logger::Manager` and the `Collector`s/`Filters` - which are easily accessible through the manager.

Also, if you give people the *ability* to add filters and emitters to leaf loggers, they *will* add filters and emitters to leaf loggers. If you add filters and emitters to leaf loggers, the data flow diagram does turn into spaghetti.

File: src/logger.cr (hunk 7 above)
[the remark this answers is not in the document]
Comment (ezrast) [reply src/logger.cr]: @straight-shoota I agree, but for now I'll be consistent with the rest of the codebase. Removing wildcard requires from the language is a fight for tomorrow. ;)

**Comment by ezrast** [comment]
^ This splits out the guts of `Logger` into `Logger::Dispatcher` ("manager" is a very non-descriptive name imo). `Logger` itself is now a thin struct. What was `Logger::ROOT` is now `Logger.default_dispatcher`. Gist updated: https://gist.github.com/ezrast/37c3377c172cd225efe0ed40f4da1e59/revisions

I think @straight-shoota has a good point about putting filters on the Emitter. Emitters may do globally-stateful things like rate limiting or database locking so it makes sense to want to use the same Emitter in two different places that may not have the same log level.

Dispatcher includes Emitter so you can chain Dispatchers together if you want some Emitters to use a separate log level:
```crystal
# This MyEmitter instance will respond only to warnings and above
Logger.default_dispatcher.emitters.push(
  Logger::Dispatcher.new(Logger::WARN, MyEmitter.new)
)
```

**Comment by RX14** [comment]
Looks good. I'll wait for @bcardiff to comment on the design before doing a full review.

**Comment by z64** [comment]
Could this get a review :)? 🏓 This is at least blocking `crystal-db` from getting out of the box logging which is sorely needed, and the rest of the driver ecosystem would immediately benefit from. (https://github.com/crystal-lang/crystal-db/pull/64#issuecomment-393257027). 

It would be great if the design here could be approved so that this PR can continue, or the work can be picked up and completed in another.

**Comment by ezrast** [comment]
Thanks for the bump, @z64.

If it would help this get reviewed I can find time to add in documentation soon. Otherwise I'm holding off on docs/tests until the design gets approval.

**Comment by Blacksmoke16** [comment]
Prob can be closed now due to https://github.com/crystal-lang/crystal/pull/8847.
